substrate/frame/revive/src/storage/meter.rs

```diff
@@ -456,18 +453,17 @@ impl<T: Config> Ext<T> for ReservingExt {
 		contract: &T::AccountId,
 		amount: &DepositOf<T>,
 		state: &ContractState<T>,
+		exec_config: &ExecConfig,
 	) -> Result<(), DispatchError> {
 		match amount {
 			Deposit::Charge(amount) | Deposit::Refund(amount) if amount.is_zero() => return Ok(()),
 			Deposit::Charge(amount) => {
-				T::Currency::transfer_and_hold(
-					&HoldReason::StorageDepositReserve.into(),
+				<Pallet<T>>::charge_deposit(
+					Some(HoldReason::StorageDepositReserve),
 					origin,
 					contract,
 					*amount,
-					Precision::Exact,
-					Preservation::Preserve,
-					Fortitude::Polite,
+					exec_config,
 				)?;
 			},
 			Deposit::Refund(amount) => {
```

Comment:
In the situation of a refund, shouldn't we this refund to the imbalance stored in pallet-transaction payment, i.e., call `T::FeeInfo::deposit_txfee` instead of refunding to the origin directly via `T::Currency::transfer_on_hold` (similarly to how end up calling `T::FeeInfo::withdraw_txfee` in the `Deposit::Charge` case above)?

Reason: if there is a big charge and an equally big refund, then the total storage deposit determined in dry running is 0 and therefore not accounted for when creating the imbalance later. Then the imbalance is not sufficient to contain the post-dispatch transaction fee and the transaction fails.

Reply:
Ahh yes good point. We assume that the refunds are available when charging. This is why we do them first.

Reply:
Fixed in my last commit.